src/core/room-lifecycle-manager.ts

```diff
@@ -1,923 +1,394 @@
+ * while maintaining the room's status.
  */
-export class RoomLifecycleManager {
-  /**
-   * The status of the room.
-   */
-  private readonly _lifecycle: DefaultRoomLifecycle;
-
-  /**
-   * The features that contribute to the room status.
-   */
-  private readonly _contributors: ContributesToRoomLifecycle[];
+export class RoomLifeCycleManager {
+  private readonly _channelManager: ChannelManager;
+  private readonly _roomLifecycle: InternalRoomLifecycle;
   private readonly _logger: Logger;
-
-  /**
-   * This mutex allows us to ensure the integrity and atomicity of operations that affect the room status, such as
-   * attaching, detaching, and releasing the room. It makes sure that we don't have multiple operations happening
-   * at once which could leave us in an inconsistent state.
-   */
-  private readonly _mtx = new Mutex();
-
-  /**
-   * A map of contributors to transient detach timeouts.
-   *
-   * If a channel enters the attaching state (as a result of a server initiated detach), we should initially
-   * consider it to be transient and not bother changing the room status.
-   */
-  private readonly _transientDetachTimeouts: Map<ContributesToRoomLifecycle, ReturnType<typeof setTimeout>>;
-
-  /**
-   * This flag indicates whether some sort of controlled operation is in progress (e.g. attaching, detaching, releasing).
-   *
-   * It is used to prevent the room status from being changed by individual channel state changes and ignore
-   * underlying channel events until we reach a consistent state.
-   */
-  private _operationInProgress = false;
-
-  /**
-   * A map of pending discontinuity events.
-   *
-   * When a discontinuity happens due to a failed resume, we don't want to surface that until the room is consistently
-   * attached again. This map allows us to queue up discontinuity events until we're ready to process them.
-   */
-  private _pendingDiscontinuityEvents: DiscontinuityEventMap = new Map();
-
-  /**
-   * A map of contributors to whether their first attach has completed.
-   *
-   * Used to control whether we should trigger discontinuity events.
-   */
-  private _firstAttachesCompleted = new Map<ContributesToRoomLifecycle, boolean>();
-
-  /**
-   * Are we in the process of releasing the room?
-   */
-  private _releaseInProgress = false;
-
-  /**
-   * Constructs a new `RoomLifecycleManager` instance.
-   * @param lifecycle The room lifecycle that manages status.
-   * @param contributors The features that contribute to the room status.
-   * @param logger An instance of the Logger.
-   * @param transientDetachTimeout The number of milliseconds to consider a detach to be "transient"
-   */
-  constructor(
-    lifecycle: DefaultRoomLifecycle,
-    contributors: ContributesToRoomLifecycle[],
-    logger: Logger,
-    transientDetachTimeout: number,
-  ) {
+  private readonly _roomId: string;
+  private readonly _eventEmitter: EventEmitter<RoomLifeCycleEvents>;
+  private _isFirstAttach: boolean; // CHA-RL13
```

Comment:
```suggestion
  private _isAttachedOnce: boolean; // CHA-RL13
```
Initial value will be `false`

Reply:
Since, variable name is explicit, there's no guessing as to whether it's set before first ATTACH or after first ATTACH.

Reply:
[d004110](https://github.com/ably/ably-chat-js/pull/505/commits/d0041104aa425104e931b758c8649ba11987bb27)